app/controllers/middleware_server_controller.rb

```diff
@@ -28,9 +28,69 @@ def listicon_image(item, _view)
     "svg/#{icon}.svg"
   end
 
+  def button
+    super unless %w(middleware_server_reload middleware_server_stop).include?(params[:pressed])
+
+    operation = params[:pressed]
+    items = identify_selected_servers
+
+    return if items.nil? # TODO: re-route somewhere?
+
+    run_server_operation(operation, items)
+
+    render :update do |page|
+      page.replace("flash_msg_div", :partial => "layouts/flash_msg")
+    end
+  end
+
   private ############################
 
   def display_name
     _('Middleware Servers')
   end
+
+  def identify_selected_servers
+    items = params[:miq_grid_checks]
+    return items unless items.nil? || items.empty?
+
+    identify_record(params[:id])
+  end
+
+  def run_server_operation(operation, items)
+    reload_done = false
+    items.split(/,/).each do |item|
+      mw_server = identify_record item
+      if mw_server.product == 'Hawkular'
+        add_flash(_(get_msg_for_op(operation, :hwkserver)))
+      else
+        mw_reload operation, mw_server # TODO: send all requests in one go instead of looping here ?
+        reload_done = true
+      end
+    end
+    add_flash(_(get_msg_for_op(operation))) if reload_done
+  end
+
+  def mw_reload(operation, mw_server)
+    mw_manager = mw_server.ext_management_system
+
+    case operation
+    when 'middleware_server_reload'
+      mw_manager.reload_middleware_server mw_server.ems_ref
+    when 'middleware_server_stop'
+      mw_manager.stop_middleware_server mw_server.ems_ref
+    else
+      raise "Unknown operation #{operation}"
+    end
+  end
+
+  def get_msg_for_op(operation, kind = :ok)
+    case operation
+    when 'middleware_server_reload'
```

Comment:
like my note above, it seems a bit redundant code duplication, the only difference is the words "reloading" versus "stopping", maintainance wise in the future it will be easier to have it more generic. what do you think?